src/utilities/interfaces.ts

```diff
@@ -0,0 +1,16 @@
+import { ErrorObject } from 'ajv';
+
+/**
```

Comment:
Could you tell me more about the design choices in the following code? In particular I am interested in the type and the global variable with the same name and the streamlined syntax for constructing the YAMLValidationError. Is this a standard pattern for TypeScript? Did you consider other approaches?

Reply:
For this, I just followed the existing pattern I found for TypeError and so I followed that:

```
interface TypeError extends Error {
}

interface TypeErrorConstructor {
 new(message?: string): TypeError;
 (message?: string): TypeError;
 readonly prototype: TypeError;
}

declare var TypeError: TypeErrorConstructor;
```

Upon further investigation I don't think we need to create a new interface; TypeError is able to handle more information such as filename, line number and error stack etc. I propose we scrap this and go back to using TypeError, but just add more detail to our exceptions. 

For example:
```
const message = 'validateRuleConfig: Invalid `rules.yaml` config file.';
throw new TypeError(message, 'rules.yaml', lineNumber, ruleConfigValidator.errors);
```

However, I have not found a way to extract the line number from ajv; we could potentially use jsonPointers and maybe report the location in the JSON input as a JSON pointer.

`const ajv = new AJV( { jsonPointers: true } );`

We could also consider using the [betterAjvErrors](https://github.com/atlassian/better-ajv-errors) library like this:

```
const output = betterAjvErrors(schema, data, ruleConfigValidator.errors);
throw new TypeError(message, 'rules.yaml', output);
```

Thoughts?

@MikeHopcroft @KatieProchilo

Reply:
I agree with TypeError vs. YAMLValidationError - I don't think the new interface gets us anything extra.

If betterAjvErrors could get us a line number in the yaml, then I think that'd be great.